llm/training/dataprep.py

```diff
@@ -0,0 +1,323 @@
+"""
+This script converts data to a format that is suitable for training. We expect inputs to be a HuggingFace dataset that has the following fields:
+input: str
+output: Optional[str]
+contexts: List[str] # contexts you want the LLM to use in the output.
+"""
+from typing import Dict, List, Any
+import copy
+
+import click
+from datasets import Dataset, Sequence, Value, Features
+from ruamel.yaml import YAML
+from transformers import PreTrainedTokenizer, AutoTokenizer
+import torch
+
+from llm.model_configs import ModelConfig
+from llm.prompt import Message, Prompt
+from llm.training.config import DataPrepConfig
+
+
+class TrainingFormatConverter:
+    def __init__(
+        self,
+        prompt: Prompt,
+        tokenizer: PreTrainedTokenizer,
+        chunksize: int,
+        ignore_index: int = -100,
+    ):
+        self.prompt = prompt
+        self.tokenizer = tokenizer
+        self.chunksize = chunksize
+        self.ignore_index = ignore_index
+
+    def __call__(self, batch: Dict[str, List]) -> Dict[str, List]:
+        prompt = self.prompt
+        tokenizer = self.tokenizer
+        chunksize = self.chunksize
+        ignore_index = self.ignore_index
+        num_rows = len(batch["input"])
+        output = {}
+        for idx in range(num_rows):
+            input_content = batch["input"][idx]
+            response_content = batch["response"][idx]
+            if "contexts" in batch:
+                contexts = batch["contexts"][idx]
+            else:
+                contexts = None
+            input_message = Message(input=input_content, response="", contexts=contexts)
+            full_message = Message(
+                input=input_content, response=response_content, contexts=contexts
+            )
+            input_prompt = prompt.render([input_message])
+            full_text = prompt.render([full_message])
```

Comment:
I've been doing this to get the two parts (should be equivalent either way)
```
message = Message(input_content, contexts=contexts, response=response)
full_text = prompt.render_message(message)
input_text = prompt.render_instruction(message.input, message.contexts, with_response_prefix=True)
```

Reply:
Also another thing to think about is how to support training over conversational data (probably a followup for later)